catalyst/contrib/data/nlp/language_modeling/dataset.py

```diff
@@ -0,0 +1,91 @@
+from typing import Iterable, Union
+
+from tqdm.auto import tqdm
+
+import torch
+import transformers
+from transformers import AutoTokenizer
+
+
+class LanguageModelingDataset(torch.utils.data.Dataset):
+    """
+    Dataset for (masked) language model task.
+    Can sort sequnces for efficient padding.
+    """
+
+    def __init__(
+        self,
+        texts: Iterable[str],
+        tokenizer: Union[
+            str, transformers.tokenization_utils.PreTrainedTokenizer
+        ],
+        max_seq_length: int = None,
+        sort: bool = True,
+        lazy: bool = False,
+    ):
+        """
+        Args:
+            texts (Iterable): Iterable object with text
+            tokenizer (str or tokenizer): pre trained
+                huggingface tokenizer or model name
+            max_seq_length (int): max sequence length to tokenize
+            sort (bool): If True then sort all sequences by length
+                for efficient padding
+            lazy (bool): If True then tokenize sequence
+                in __getitem__ method
+                else will tokenize in __init__ also
+                if set to true sorting is unavialible
+        """
+        if isinstance(tokenizer, str):
+            self.tokenizer = AutoTokenizer.from_pretrained(tokenizer)
+        elif isinstance(
+            tokenizer, transformers.tokenization_utils.PreTrainedTokenizer
+        ):
+            self.tokenizer = tokenizer
+        else:
+            raise TypeError(
+                "tokenizer argument should be a model name"
+                + " or huggingface PreTrainedTokenizer"
+            )
+
+        self.max_seq_length = max_seq_length
+
+        self.lazy = lazy
+
+        if lazy:
+            self.texts = texts
+
+        if not lazy:
+            pbar = tqdm(texts, desc="tokenizing texts")
+            self.encoded = [
+                self.tokenizer.encode(text, max_len=max_seq_length)
+                for text in pbar
+            ]
+            if sort:
+                self.encoded.sort(key=len)
+
+        elif sort:
+            raise Exception(
+                "lazy is set to True so we can't sort"
+                " sequences by length.\n"
+                "You should set sort=False if lazy=True"
+                " if you want to encode text in __get_item__ function"
+            )
+        self.length = len(texts)
```

Comment:
btw, could we raise exception, or maybe assert in the beginning of initialisation?
currently, we could get to the situation, when we have encoded some dataset (big one) and then caught and exception - that's not cool :)